### PR TITLE
Add Managed Service Identity (MSI) support to VM Scale Sets

### PR DESCRIPTION
**Pull request by day-jeff**
This builds on earlier work enabling MSI for VMs. Adds the identity attribute and principal_id property to the VM scale set resource.

### CODE REVIEW

File: azurerm/resource_arm_virtual_machine_scale_set.go [hunk 1]
```diff
@@ -803,6 +835,22 @@ func resourceArmVirtualMachineScaleSetDelete(d *schema.ResourceData, meta interf
 	return nil
 }
 
+func flattenAzureRmVirtualMachineScaleSetIdentity(identity *compute.VirtualMachineScaleSetIdentity) []interface{} {
+	log.Printf("[DEBUG] entered flattenAzureRmVirtualMachineScaleSetIdentity")
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_virtual_machine_scale_set.go]: can we remove this debug line?

Comment (dawg89) [reply azurerm/resource_arm_virtual_machine_scale_set.go]: Yep, will do.

File: azurerm/resource_arm_virtual_machine_scale_set.go [hunk 2]
```diff
@@ -37,6 +37,29 @@ func resourceArmVirtualMachineScaleSet() *schema.Resource {
 
 			"zones": zonesSchema(),
 
+			"identity": {
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_virtual_machine_scale_set.go]: can we add an acceptance test covering this use-case?

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @dawg89 

Thanks for this PR - I've taken a look through and left some comments inline but this is off to a good start; if we can fix up the minor issues and add an acceptance test then this should be good to merge :)

Thanks!

File: azurerm/resource_arm_virtual_machine_scale_set.go (hunk 2 above)
Comment (dawg89) [reply azurerm/resource_arm_virtual_machine_scale_set.go]: Done

File: website/docs/r/virtual_machine_scale_set.html.markdown [hunk 3]
```diff
@@ -270,6 +270,41 @@ The following arguments are supported:
 * `tier` - (Optional) Specifies the tier of virtual machines in a scale set. Possible values, `standard` or `basic`.
 * `capacity` - (Required) Specifies the number of virtual machines in the scale set.
 
+`identity` supports the following:
+
+* `type` - (Required) Specifies the identity type of the virtual machine. The only allowable value is `SystemAssigned`. To enable Managed Service Identity (MSI) on all machines in the scale set, an extension with the type "ManagedIdentityExtensionForWindows" or "ManagedIdentityExtensionForLinux" must also be added. The scale set's Service Principal ID (SPN) can be retrieved after the scale set has been created.
```
[the remark this answers is not in the document]
Comment (dawg89) [reply website/docs/r/virtual_machine_scale_set.html.markdown]: fixed

File: azurerm/resource_arm_virtual_machine_scale_set.go [hunk 4]
```diff
@@ -639,6 +662,13 @@ func resourceArmVirtualMachineScaleSetCreate(d *schema.ResourceData, meta interf
 		Zones: zones,
 	}
 
+	if _, ok := d.GetOk("identity"); ok {
+		log.Printf("[DEBUG] attempting to enable MSI")
```
[the remark this answers is not in the document]
Comment (dawg89) [reply azurerm/resource_arm_virtual_machine_scale_set.go]: done

File: azurerm/resource_arm_virtual_machine_scale_set.go [hunk 5]
```diff
@@ -639,6 +662,13 @@ func resourceArmVirtualMachineScaleSetCreate(d *schema.ResourceData, meta interf
...
+		scaleSetParams.Identity = expandAzureRmVirtualMachineScaleSetIdentity(d)
+	} else {
+		log.Printf("[DEBUG] MSI failed")
```
[the remark this answers is not in the document]
Comment (dawg89) [reply azurerm/resource_arm_virtual_machine_scale_set.go]: done

File: website/docs/r/virtual_machine_scale_set.html.markdown [hunk 6]
```diff
@@ -270,6 +270,41 @@ The following arguments are supported:
...
+  }
+
+  extension {
+    name                       = "MSILinuxExtension"
+    publisher                  = "Microsoft.ManagedIdentity"
+    type                       = "ManagedIdentityExtensionForLinux"
+    type_handler_version       = "1.0"
+    auto_upgrade_minor_version = true
+    settings                   = "{\"port\": 50342}"
+    protected_settings         = "{}"
```
[the remark this answers is not in the document]
Comment (dawg89) [reply website/docs/r/virtual_machine_scale_set.html.markdown]: Tested and removed both auto_upgrade_minor_version & protected_settings.

**Review by day-jeff — COMMENTED** [review]
Great review, Tom. Thanks!

**Comment by dawg89** [comment]
All review comments addressed. Ready for round 2. :)

File: azurerm/resource_arm_virtual_machine_scale_set.go [hunk 7]
```diff
@@ -701,6 +728,8 @@ func resourceArmVirtualMachineScaleSetRead(d *schema.ResourceData, meta interfac
 		return fmt.Errorf("[DEBUG] Error setting Virtual Machine Scale Set Sku error: %#v", err)
 	}
 
+	d.Set("identity", flattenAzureRmVirtualMachineScaleSetIdentity(resp.Identity))
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_virtual_machine_scale_set.go]: can we change this to:

```
if err := d.Set("identity", flattenAzureRmVirtualMachineScaleSetIdentity(resp.Identity)); err != nil {
  return fmt.Error("Error flattening `identity`: %+v", err)
}
```

this allows any bugs (such as the schema not matching) to be caught

File: azurerm/resource_arm_virtual_machine_scale_set_test.go [hunk 8]
```diff
@@ -418,6 +418,26 @@ func TestAccAzureRMVirtualMachineScaleSet_overprovision(t *testing.T) {
...
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testCheckAzureRMVirtualMachineScaleSetDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: config,
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMVirtualMachineScaleSetExists(resourceName),
+					testCheckAzureRMVirtualMachineScaleSetMSI(resourceName),
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_virtual_machine_scale_set_test.go]: can we remove this function (which checks the API) in favour of checking the value stored in the state (which users will consume)? we can do this via:

```
resources.TestCheckResourceAttrSet(resourceName, "identity.0.principal_id"),
```

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @dawg89 

Thanks for pushing those updates (and apologies for the delayed re-review here!). I've noted two minor comments but this otherwise LGTM 👍. If we can get those sorted then this should be good to merge :)

Thanks!

File: azurerm/resource_arm_virtual_machine_scale_set.go (hunk 7 above)
Comment (dawg89) [reply azurerm/resource_arm_virtual_machine_scale_set.go]: Done

File: azurerm/resource_arm_virtual_machine_scale_set_test.go (hunk 8 above)
Comment (dawg89) [reply azurerm/resource_arm_virtual_machine_scale_set_test.go]: Done

**Comment by dawg89** [comment]
More good changes, Tom. Thx for the reviews. 

--Jeff

**Review by tombuildsstuff — APPROVED** [review]
hey @dawg89 

Thanks for pushing the latest changes - this now LGTM 👍 I'll kick off the test suite now :)

Thanks!

**Comment by tombuildsstuff** [comment]
@dawg89 just to let you know that the tests picked up a bug in the tests, so I've pushed a commit to fix this: https://github.com/terraform-providers/terraform-provider-azurerm/pull/1018/commits/03a9c31582e32caa7e50ca8e85294e9f7af74a68 - I hope you don't mind :)

**Comment by tombuildsstuff** [comment]
All VM tests before the previous commit:

screen shot 2018-04-04 at 13 36 06

The final commit fixes the broken test:

screen shot 2018-04-04 at 13 35 59

Good to merge 👍

**Comment by herrkunstler** [comment]
According to https://docs.microsoft.com/en-us/azure/active-directory/managed-service-identity/overview, "MSI VM extension endpoint: http://localhost:50342/oauth2/token (to be deprecated)." Since the extension endpoint is being deprecated, will this provider be updated to support enabling MSI without the extension (see also https://docs.microsoft.com/en-us/azure/active-directory/managed-service-identity/tutorial-linux-vm-access-arm)?

**Pull request merged**